Setup Github action: Run benchmarks and publish results

mbatoul opened the pull request:
- Run benchmarks when commits are pushed on this branch. Next: change for `main`.
- Generate HTML file from notebook with `nbconvert`.
- Publish results on branch `gh-pages`.

For now, benchmarks run in one hour or so:
![image](https://user-images.githubusercontent.com/36322423/118336782-a78ce680-b512-11eb-9999-70efd8502b2d.png)

mbatoul committed:
Add workflow to publish benchmark results on gh-pages

mbatoul committed:
Fix

mbatoul committed:
Add README.md

jjerphan commented:
Is this branch ready for review, @mbatoul?

mbatoul commented:
If you're happy with the [published results](https://scikit-learn-inria-fondation.github.io/pdist_aggregation/), it's ready for review yes!

jjerphan commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,68 @@
+          export NUM_CORES=`python -c "import joblib; print(joblib.cpu_count())"`
+          export MKL_NUM_THREADS=$NUM_CORES
+          export OMP_NUM_THREADS=$NUM_CORES
+          export OPENBLAS_NUM_THREADS=$NUM_CORES
```
We need to avoid threads over-subscription when using BLAS/MKL within code parallelised by OpenMP, which must happen here.

I think of adding this in another PR improving benchmarks which would introduce `threadpoolctl` to do so. For the time being we can solve it setting the environment variables accordingly.
```suggestion
          export OMP_NUM_THREADS=$NUM_CORES
          # Avoid threads over-subscription in parallel implementation 
          # TODO: remove once threadpoolctl is set.
          export MKL_NUM_THREADS=1
          export OPENBLAS_NUM_THREADS=1
```

jjerphan committed:
Tidy-up notebook

Remove previous runs output.
Add threads configuration on figures.
List specifications.

jjerphan reviewed (changes requested):
Thanks @mbatoul!

I've added https://github.com/scikit-learn-inria-fondation/pdist_aggregation/pull/8/commits/bb3857641b486316ed80a5ee123e417823b1205f for small improvements on the notebook as well as a suggestion to fix threads over-subscription which might be present.

mbatoul committed:
Set MKL_NUM_THREADS=1 and OPENBLAS_NUM_THREADS=1 to avoid oversubscription

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jjerphan committed:
Import missing submodule

jjerphan commented:
Thanks @mbatoul.
